o1vm/src/keccak/trace.rs

```diff
@@ -7,118 +7,106 @@ use strum::IntoEnumIterator;
     },
-    trace::{DecomposableTracer, DecomposedTrace},
+    trace::{DecomposableTracer, DecomposedTrace, Trace, Tracer},
 };
 
 use super::folding::KeccakConfig;
 
+/// A Keccak instruction trace
+pub type KeccakTrace = Trace<N_ZKVM_KECCAK_COLS, KeccakConfig>;
 /// The Keccak circuit trace
-pub type KeccakTrace = DecomposedTrace<
-    N_ZKVM_KECCAK_COLS,
-    N_ZKVM_KECCAK_REL_COLS,
-    N_ZKVM_KECCAK_SEL_COLS,
-    KeccakConfig,
->;
+pub type DecomposedKeccakTrace = DecomposedTrace<N_ZKVM_KECCAK_COLS, KeccakConfig>;
 
-impl
-    DecomposableTracer<
-        N_ZKVM_KECCAK_COLS,
-        N_ZKVM_KECCAK_REL_COLS,
-        N_ZKVM_KECCAK_SEL_COLS,
-        KeccakConfig,
-        KeccakEnv<ScalarField<KeccakConfig>>,
-    > for KeccakTrace
-{
-    fn new(domain_size: usize, _env: &mut KeccakEnv<ScalarField<KeccakConfig>>) -> Self {
+impl DecomposableTracer<KeccakEnv<ScalarField<KeccakConfig>>> for DecomposedKeccakTrace {
+    fn new(domain_size: usize, env: &mut KeccakEnv<ScalarField<KeccakConfig>>) -> Self {
         let mut circuit = Self {
             domain_size,
-            witness: BTreeMap::new(),
-            constraints: Default::default(),
-            lookups: Default::default(),
+            trace: BTreeMap::new(),
         };
-
-        for opcode in Steps::iter().flat_map(|step| step.into_iter()) {
-            circuit.witness.insert(
-                opcode,
-                Witness {
-                    cols: Box::new(std::array::from_fn(|_| Vec::with_capacity(domain_size))),
-                },
-            );
+        for step in Steps::iter().flat_map(|step| step.into_iter()) {
             circuit
-                .constraints
-                .insert(opcode, KeccakEnv::constraints_of(opcode));
-            circuit
-                .lookups
-                .insert(opcode, KeccakEnv::lookups_of(opcode));
+                .trace
+                .insert(step, KeccakTrace::init(domain_size, step, env));
         }
         circuit
     }
 
+    fn pad_witnesses(&mut self) {
+        for opcode in Steps::iter().flat_map(|opcode| opcode.into_iter()) {
+            if self.in_circuit(opcode) {
+                self.trace.get_mut(&opcode).unwrap().pad_dummy(());
+            }
+        }
+    }
+}
+
+impl Tracer<N_ZKVM_KECCAK_REL_COLS, KeccakConfig, KeccakEnv<ScalarField<KeccakConfig>>>
+    for KeccakTrace
+{
+    type Selector = ();
+
+    fn init(
+        domain_size: usize,
+        selector: Steps,
+        _env: &mut KeccakEnv<ScalarField<KeccakConfig>>,
+    ) -> Self {
+        // Make sure we are using the same round number to refer to round steps
+        let step = standardize(selector);
+        Self {
+            domain_size,
+            witness: Witness {
+                cols: Box::new(std::array::from_fn(|_| Vec::with_capacity(domain_size))),
+            },
+            constraints: KeccakEnv::constraints_of(step),
+            lookups: KeccakEnv::lookups_of(step),
+        }
+    }
+
     fn push_row(
```

Comment:
Wondering if we can still improve this duplication of code. We might want to discuss again what @volhovm had in mind for these "shared" implementations. For each structure implementing the tracer trait, we will have nearly the same implementation.